src/main/scala/mesosphere/marathon/core/storage/zookeeper/ZooKeeperPersistenceStore.scala

```diff
@@ -0,0 +1,274 @@
+  private[this] val transactionMetric        = Metrics.counter(ServiceMetric, getClass, "transaction")
+  private[this] val transactionOpCountMetric = Metrics.counter(ServiceMetric, getClass, "transactionOpCount")
+
+
+  // Helper stages for logging and metrics
+  def logNode(prefix: String): Flow[Node, Node, NotUsed]                = Flow[Node].map{ n => logger.debug(s"$prefix${n.path}"); n }
+  def logPath(prefix: String): Flow[String, String, NotUsed]            = Flow[String].map{ p => logger.debug(s"$prefix${p}"); p }
+  def metric[T](counter: Counter, times: Long = 1): Flow[T, T, NotUsed] = Flow[T].map{ e => counter.increment(times); e}
+  // format: ON
+
+  /**
+    * A Flow for saving nodes to the store. It takes a stream of nodes and returns a stream of node keys
+    * that were successfully stored.
+    *
+    * By default persistent nodes with empty ACLs are created. If the path is nested parent nodes will be also
+    * created if they don't already exist. If there is already a node with the same path a [[org.apache.zookeeper.KeeperException.NodeExistsException]]
+    * is thrown.
+    *
+    * @return
+    */
+  override def createFlow: Flow[Node, String, NotUsed] =
+    Flow[Node]
+      .via(logNode("Creating a node at "))
+      .via(metric(createMetric))
+      // `groupBy(path.hashCode % parallelism)` makes sure that updates to the same path always land in the same
+      // sub-stream, thus keeping the order of writes to the same path, even with parallelism > 1
+      .groupBy(parallelism, node => Math.abs(node.path.hashCode) % parallelism)
+      .mapAsync(parallelism)(node =>
+        factory
+          .create()
+          .forPath(node.path, node.data.toArray)
+          .toScala)
+      .mergeSubstreams
+
+  override def create(node: Node): Future[String] = {
+    logger.debug(s"Creating a node at ${node.path}")
+    createMetric.increment()
+    factory
+      .create()
+      .forPath(node.path, node.data.toArray)
+      .toScala
+  }
+
+  /**
+    * A Flow for reading nodes from the store. It takes a stream of node paths and returns a stream of Try[Node] elements.
+    * It's a Success[Node] for existing nodes or Failure(e) for non-existing nodes where e is an instance of
+    * [[org.apache.zookeeper.KeeperException.NoNodeException]]. The exception contains the path to the node which
+    * simplifies handling failed reads. For other exceptions stream is completed with a failure.
+    *
+    * @return
+    */
+  override def readFlow: Flow[String, Try[Node], NotUsed] =
+    Flow[String]
+      .via(logPath("Reading a node at "))
+      .via(metric(readMetric))
+      .mapAsync(parallelism)(path =>
+        factory
+          .getData()
+          .forPath(path)
+          .toScala
+          .map(bytes => Try(Node(path, ByteString(bytes))))
+          .recover {
+            case e: NoNodeException => Failure(e) // re-throw exception in all other cases
+          }
```

Comment:
Same as above. There is some code duplication and I think the flows should reuse the simple methods that return the futures.